addons/bus/static/src/workers/websocket_worker.js

```diff
@@ -0,0 +1,283 @@
+/** @odoo-module **/
+
+/**
+ * Type of events that can be sent from the worker to its clients.
+ *
+ * @typedef { 'connect' | 'reconnect' | 'disconnect' | 'reconnecting' |
+ * 'server_error' | 'notification' } WorkerEvent
+ */
+
+/**
+ * Type of action that can be sent from the client to the worker.
+ *
+ * @typedef {'add_channel' | 'delete_channel' | 'force_update_channels'
+ * | 'send' | 'leave' } WorkerAction
+ */
+
+/**
+ * This class regroups the logic necessary in order for the
+ * SharedWorker/Worker to work. Indeed, Safari and some minor browsers
+ * does not support SharedWorker. In order to solve this issue, a Worker
+ * is used in this case. The logic is almost the same than the one used
+ * for SharedWorker and this class implements it.
+ */
+class WebsocketWorker {
+    constructor(websocketURL, websocketClass = WebSocket, setTimeoutFn = setTimeout) {
```

Comment:
It's not good to have complex API caused by mocking in tests.
In practice only `websocketURL` is necessary.

The test things should not make this code more complex than it should.

I suppose the problem is that worker has no `window`, but you want mocked things in tests that are in main thread rather than in actually a worker.

The proper way would have a very small part of code that is just technical websocket (communication in/out, essentially), and then have the rest being defined in a proper model, e.g. `BusNotificationAgent` of something like that.